checker/src/org/checkerframework/checker/regex/RegexQualifiedTransfer.java

```diff
@@ -45,82 +59,218 @@ public RegexQualifiedTransfer(QualAnalysis<QualParams<Regex>> analysis) {
+        // Make mat be @Regex(constant)
+        TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> res;
+        res = super.visitLessThanOrEqual(n, in);
+        res = handleMatcherGroupCount(n.getRightOperand(), n.getLeftOperand(), true, in, res);
+        return res;
+    }
+
+    @Override
+    public
+    TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>>
+    visitGreaterThan(GreaterThanNode n,
+            TransferInput<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> in) {
+        // Look for: mat.groupCount() > constant
+        // Make mat be @Regex(constant + 1)
+        TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> res;
+        res = super.visitGreaterThan(n, in);
+        res = handleMatcherGroupCount(n.getLeftOperand(), n.getRightOperand(), false, in, res);
+        return res;
+    }
+
+    @Override
+    public
+    TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>>
+    visitGreaterThanOrEqual(GreaterThanOrEqualNode n,
+            TransferInput<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> in) {
+        // Look for: mat.groupCount() >= constant
+        // Make mat be @Regex(constant)
+        TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> res;
+        res = super.visitGreaterThanOrEqual(n, in); 
+        res = handleMatcherGroupCount(n.getLeftOperand(), n.getRightOperand(), true, in, res);
+        return res;
+    }
+
+    /**
+     * See whether possibleMatcher is a call of groupCount on a Matcher and
+     * possibleConstant is a constant.
+     * If so, annotate the matcher as
+     * constant + 1 if !isAlsoEqual
+     * constant if isAlsoEqual
+     *
+     * @param possibleMatcher the Node that might be a call of Matcher.groupCount()
+     * @param possibleConstant the Node that might be a constant
+     * @param isAlsoEqual whether the comparison operation is strict or reflexive
+     * @param in the input TransferResult
+     * @return the possibly refined output TransferResult
+     */
+    private
+    TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>>
+    handleMatcherGroupCount(
+            Node possibleMatcher, Node possibleConstant, boolean isAlsoEqual,
+            TransferInput<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> in,
+            TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> resultIn) {
+        if (!(possibleMatcher instanceof MethodInvocationNode)) {
+            return resultIn;
+        }
+        if (!(possibleConstant instanceof IntegerLiteralNode)) {
+            return resultIn;
+        }
+
+        MethodInvocationNode min = (MethodInvocationNode) possibleMatcher;
+        MethodAccessNode target = min.getTarget();
+        ExecutableElement method = target.getMethod();
+
+        if (!ElementUtils.matchesElement(method, GROUP_COUNT_METHOD_NAME)) {
```

Comment:
This does not seem to be checking the class name or number of arguments, but I think it should.  There are groupCount methods in both MatchResult and Matcher -- is it intended to apply to both?

Is this a known deficiency that is intended to be corrected once we use annotations to indicate special behavior, rather than hard-coding method names in source code?  (I can't find the relevant issue or pull request offhand.)  If so, I think it would be worthwhile to apply that fix to just this file, if possible:  we can improve the code as we edit every file rather than waiting to do it to every class.